Make `skip_history_when_saving` work when creating a new object (#1262)
jazzband/django-simple-history: docs/querying_history.rst

    <Poll: Poll object as of 2010-10-25 18:04:13.814128>

> Save Without Creating Historical Records

Reviewer: Most titles don't use title casing in the file. I think you should reconsider this choice.

Author: Ah yes, I agree 👌 I was planning on polishing the docs in a later PR, which includes title-casing all the headings - to match both the main headings in the docs sidebar and various subheadings on some of the pages.

Any opinions on whether we should make title casing or normal casing the norm for the subheadings, btw?

Reviewer: I prefer the normal casing for sentence titles here, but it's not something I'd argue to change an existing set of documentation over to.

----------------------------------------

If you want to save model objects without triggering the creation of any historical
records, you can set a ``skip_history_when_saving`` attribute to ``True``

> on each object before saving - for example like this:

Reviewer: Personally, I think the previous version is better. It's fewer words and you have a code example to show the how.

Author: Good point on already having a code example 😅 I'll push a slightly shorter version 🙂 I do prefer being specific about "triggering the creation", though.

.. code-block:: python

    poll.skip_history_when_saving = True
    poll.save()
    # We recommend deleting the attribute afterward
    del poll.skip_history_when_saving

> This also works when creating an object
> (albeit only when calling the model's standard Python constructor followed by saving):

Reviewer: ```suggestion
This also works when creating an object, but only when calling the ``save`` method:
```

I think it simplifies the language a bit. I'm not tied particularly strongly to it.

Author: Sure! `save()` is what triggers our signal handlers, after all 👌

.. code-block:: python

>     # Note that `Poll.objects.create()` is not called

Reviewer: Maybe we put this after the code example?

> Historical records will always be created using the ``create`` manager method.

Reviewer: Also I bet it's possible for us to manage this by overriding the ``create`` method on the manager class and accepting ``skip_history_when_saving`` as a kwarg.

Author: > Maybe we put this after the code example?
> > Historical records will always be created using the ``create`` manager method.

Sure, I can place it inside a `.. note::` :)

Reviewer: > You think we should do that in this PR?

Definitely no :smile: A context manager sounds like a fantastic idea. Dang, that's clever!

Author: Alright :) I agree, though credits for the idea goes to https://github.com/jazzband/django-simple-history/issues/642#issuecomment-922110914 😄

    poll = Poll(question="Why?")
    poll.skip_history_when_saving = True
    poll.save()
    del poll.skip_history_when_saving

Alternatively, call the ``save_without_historical_record()`` method on each object
instead of ``save()``.
This method is automatically added to a model when registering it for history-tracking
(i.e. defining a ``HistoricalRecords``  manager field on the model),
and it looks like this:

.. code-block:: python

    def save_without_historical_record(self, *args, **kwargs):
        self.skip_history_when_saving = True
        try:
            ret = self.save(*args, **kwargs)
        finally:
            del self.skip_history_when_saving
        return ret

> As a final option, you can disable the creation of historical records for *all* models

Reviewer: ```suggestion
Or disable the creation of historical records for *all* models
```

I think the previous version was a bit too wordy for technical documentation.

Author: Fair, I'll add it :)
